compiler/rustc_codegen_llvm/src/builder.rs

```diff
@@ -24,17 +25,255 @@ use rustc_span::Span;
+        let mut fnc_args: Vec<&Value> = vec![];
+        fnc_args.reserve(param_num);
+        llvm::LLVMGetParams(fnc, fnc_args.as_mut_ptr());
+        fnc_args.set_len(param_num);
+        fnc_args
+    }
+}
+
+// The lowering of one `#[autodiff]` macro happens in multiple steps.
+// First we transalte generate a new dummy function, who's llvm-ir we now have as outer_fn.
+// We kept track of the original function to which the `#[autodiff]` macro was applied to, which we
+// now have as fn_to_diff. In our current implementation, we use the enzyme pass to carry out the
+// differentiation, following naming and calling conventions documented here: <https://enzyme.mit.edu/getting_started/CallingConvention/>
+//
+// Our `outer_fn` had some dummy code inserted at higher levels, so we first remove most of the
+// existing body. We then insert an `__enzyme_<autodiff/fwddiff>_<unique_id>` call, which the pass
+// will then pick up. FIXME(ZuseZ4): We will later want to upstream safety checks to the `outer_fn`,
+// in order to cover some assumptions of enzyme/autodiff, which could lead to UB otherwise.
+pub(crate) fn generate_enzyme_call<'ll>(
+    llmod: &'ll llvm::Module,
+    llcx: &'ll llvm::Context,
+    fn_to_diff: &'ll Value,
+    outer_fn: &'ll Value,
+    attrs: AutoDiffAttrs,
+) {
+    let inputs = attrs.input_activity;
+    let output = attrs.ret_activity;
+
+    // We have to pick the name depending on whether we want forward or reverse mode autodiff.
+    // FIXME(ZuseZ4): The new pass based approach should not need the *First method anymore, since
+    // it will handle higher-order derivatives correctly automatically (in theory). Currently
+    // higher-order derivatives fail, so we should debug that before adjusting this code.
+    let mut ad_name: String = match attrs.mode {
+        DiffMode::Forward => "__enzyme_fwddiff",
+        DiffMode::Reverse => "__enzyme_autodiff",
+        DiffMode::ForwardFirst => "__enzyme_fwddiff",
+        DiffMode::ReverseFirst => "__enzyme_autodiff",
+        _ => panic!("logic bug in autodiff, unrecognized mode"),
+    }
+    .to_string();
+
+    // add outer_fn name to ad_name to make it unique, in case users apply autodiff to multiple
+    // functions. Unwrap will only panic, if LLVM gave us an invalid string.
+    let name = llvm::get_value_name(outer_fn);
+    let outer_fn_name = std::ffi::CStr::from_bytes_with_nul(name).unwrap().to_str().unwrap();
+    ad_name.push_str(outer_fn_name.to_string().as_str());
+
+    // Assuming that our fn_to_diff is the fnc square, want to generate the following llvm-ir, which
+    // would allow the enzyme pass to generate a function body for `__enzyme_autodiff_square`
+    //
+    // declare double @__enzyme_autodiff_square(...)
+    //
+    // define double @dsquare(double %x) {
+    // entry:
+    //   %0 = tail call double (...) @__enzyme_autodiff_square(double (double)* nonnull @square, double %x)
+    //   ret double %0
+    // }
```

Comment:
This comment is useful bit it would be useful to clarify which of `dsquare`, `square`, and `__enzyme_autodiff_square` are which parameters to this function or locals in it.

Reply:
I expanded the example and description, does it help?